### PR TITLE
ISSUE #249 Fixing inherited file handles

### PR DESCRIPTION
**Pull request by bexelbie**
On Mac and Linux the DNS server spawn code automatically inherited
filehandles for STDIN, STDOUT, and STDERR.  This fix closes those
filehandles and sets all other filehandles to not explicitly redirected
to uninherited.

This is already the default for Windows.

This fix is thanks to @dustymabe and his research on a similiar issue
for vagrant-sshfs
https://github.com/dustymabe/vagrant-sshfs/issues/41

### CODE REVIEW

**Comment by bexelbie** [comment]
This also fixes https://github.com/projectatomic/adb-atomic-developer-bundle/issues/513

File: lib/landrush/server.rb [hunk 1]
```diff
@@ -67,16 +67,38 @@ def self.upstream
...
+        # the shell creating the process is closed.
         # See https://github.com/vagrant-landrush/landrush/issues/199
+        #
+        # Note to the Future: Windows does not have a
+        # file handle inheritance issue like Linux and Mac (see:
+        # https://github.com/vagrant-landrush/landrush/issues/249)
+        #
+        # On windows, if no filehandle is passed then no files get
+        # inherited by default, but if any filehandle is passed to
+        # a spawned process (as we are doing) then all files that are
```
Comment (dustymabe) [inline lib/landrush/server.rb]: `as we are doing` - what file handles are we passing to the spawned process on windows?

File: lib/landrush/server.rb [hunk 2]
```diff
@@ -67,16 +67,38 @@ def self.upstream
...
       else
-        pid = spawn('ruby', __FILE__, port.to_s, working_dir.to_s, :chdir => working_dir.to_path, :pgroup => true)
+        # Fix https://github.com/vagrant-landrush/landrush/issues/249)
+        # by turning of filehandle inheritance with :close_others => true
+        # and by explicitly closing STDIN, STDOUT, and STDERR
+
+        pid = spawn('ruby', __FILE__, port.to_s, working_dir.to_s,
+                    :in => :close, :out => :close, :err => :close,
+                    :close_others => true, :chdir => working_dir.to_path,
+                    :pgroup => true)
```
Comment (dustymabe) [inline lib/landrush/server.rb]: is this a temporary solution until we decide to send the stdout/stderr to a log file in the future that can be inspected later during debug?

Comment (bexelbie) [inline lib/landrush/server.rb]: We have a separate logging file handle opened by the server.  As far as I can tell this handles are never used.  @hferentschik blessed this in Telegram (he is on vacation)

File: lib/landrush/server.rb (hunk 1 above)
Comment (bexelbie) [inline lib/landrush/server.rb]: I will fix this grammar :)

**Comment by budhrg** [comment]
Verified the patch by running the [VagrantLauncher.java](https://gist.github.com/budhrg/648ad3e741ea2cc9ac15be8dccd221dc) program as
```
$ java VagrantLauncher <which vagrant> <Dir path of Vagrantfile> <user name> <password>
```

Found that the stdout and stderr stream thread are closed

```
ERROR>
ERROR>DEBUG ssh: stdout: Successfully started and provisioned VM with 2 cores and 3072 MB of memory.
ERROR>
ERROR> INFO interface: info: Successfully started and provisioned VM with 2 cores and 3072 MB of memory.
ERROR> INFO interface: info: ==> default: Successfully started and provisioned VM with 2 cores and 3072 MB of memory.
OUTPUT>==> default: Successfully started and provisioned VM with 2 cores and 3072 MB of memory.
ERROR>DEBUG ssh: stdout: To modify the number of cores and/or available memory set the environment variables
ERROR>VM_CPU respectively VM_MEMORY.
ERROR>
ERROR>You can now access the OpenShift console on: https://openshift.cdk:8443/console
ERROR>
ERROR> INFO interface: info: To modify the number of cores and/or available memory set the environment variables
ERROR>VM_CPU respectively VM_MEMORY.
ERROR>
ERROR>You can now access the OpenShift console on: https://openshift.cdk:8443/console
ERROR> INFO interface: info: ==> default: To modify the number of cores and/or available memory set the environment variables
ERROR>==> default: VM_CPU respectively VM_MEMORY.
ERROR>==> default: 
ERROR>==> default: You can now access the OpenShift console on: https://openshift.cdk:8443/console
OUTPUT>==> default: To modify the number of cores and/or available memory set the environment variables
OUTPUT>==> default: VM_CPU respectively VM_MEMORY.
OUTPUT>==> default: 
OUTPUT>==> default: You can now access the OpenShift console on: https://openshift.cdk:8443/console
ERROR>DEBUG ssh: stdout: 
ERROR>
ERROR>DEBUG ssh: stdout: To use OpenShift CLI, run:
ERROR>
ERROR> INFO interface: info: To use OpenShift CLI, run:
ERROR> INFO interface: info: ==> default: To use OpenShift CLI, run:
OUTPUT>==> default: To use OpenShift CLI, run:
ERROR>DEBUG ssh: stdout: $ vagrant ssh
ERROR>$ oc login
ERROR>
ERROR> INFO interface: info: $ vagrant ssh
ERROR>$ oc login
ERROR> INFO interface: info: ==> default: $ vagrant ssh
ERROR>==> default: $ oc login
OUTPUT>==> default: $ vagrant ssh
OUTPUT>==> default: $ oc login
ERROR>DEBUG ssh: stdout: 
ERROR>
ERROR>DEBUG ssh: stdout: Configured users are (<username>/<password>):
ERROR>
ERROR> INFO interface: info: Configured users are (<username>/<password>):
ERROR> INFO interface: info: ==> default: Configured users are (<username>/<password>):
OUTPUT>==> default: Configured users are (<username>/<password>):
ERROR>DEBUG ssh: stdout: openshift-dev/devel
ERROR>
ERROR> INFO interface: info: openshift-dev/devel
ERROR> INFO interface: info: ==> default: openshift-dev/devel
OUTPUT>==> default: openshift-dev/devel
ERROR>DEBUG ssh: stdout: admin/admin
ERROR>
ERROR> INFO interface: info: admin/admin
ERROR> INFO interface: info: ==> default: admin/admin
OUTPUT>==> default: admin/admin
ERROR>DEBUG ssh: stdout: 
ERROR>
ERROR>DEBUG ssh: stdout: If you have the oc client library on your host, you can also login from your host.
ERROR>
ERROR> INFO interface: info: If you have the oc client library on your host, you can also login from your host.
ERROR> INFO interface: info: ==> default: If you have the oc client library on your host, you can also login from your host.
OUTPUT>==> default: If you have the oc client library on your host, you can also login from your host.
ERROR>DEBUG ssh: stdout: 
ERROR>
ERROR>DEBUG ssh: Exit status: 0
ERROR> INFO warden: Calling OUT action: #<Proc:0x00000001e53d18@/opt/vagrant/embedded/gems/gems/vagrant-1.8.4/lib/vagrant/action/warden.rb:94 (lambda)>
ERROR> INFO warden: Calling OUT action: #<Vagrant::Action::Builtin::Provision:0x000000034fa238>
ERROR> INFO warden: Calling OUT action: #<VagrantPlugins::ProviderVirtualBox::Action::ClearForwardedPorts:0x000000034fa288>
ERROR> INFO warden: Calling OUT action: #<VagrantPlugins::ProviderVirtualBox::Action::SetName:0x0000000355d1a8>
ERROR> INFO warden: Calling OUT action: #<VagrantPlugins::ProviderVirtualBox::Action::CleanMachineFolder:0x0000000355d1d0>
ERROR> INFO warden: Calling OUT action: #<VagrantPlugins::ProviderVirtualBox::Action::CheckAccessible:0x0000000355d1f8>
ERROR> INFO warden: Calling OUT action: #<Proc:0x000000037e7388@/opt/vagrant/embedded/gems/gems/vagrant-1.8.4/lib/vagrant/action/warden.rb:94 (lambda)>
ERROR> INFO warden: Calling OUT action: #<Vagrant::Action::Builtin::Call:0x00000003b26670>
ERROR> INFO warden: Calling OUT action: #<Proc:0x000000039e6ad0@/opt/vagrant/embedded/gems/gems/vagrant-1.8.4/lib/vagrant/action/warden.rb:94 (lambda)>
ERROR> INFO warden: Calling OUT action: #<Vagrant::Action::Builtin::Call:0x00000001e52670>
ERROR> INFO warden: Calling OUT action: #<Proc:0x0000000383dc88@/opt/vagrant/embedded/gems/gems/vagrant-1.8.4/lib/vagrant/action/warden.rb:94 (lambda)>
ERROR> INFO warden: Calling OUT action: #<Vagrant::Action::Builtin::Call:0x0000000381b1b0>
ERROR> INFO warden: Calling OUT action: #<Vagrant::Action::Builtin::BoxCheckOutdated:0x00000003894d30>
ERROR> INFO warden: Calling OUT action: #<Vagrant::Action::Builtin::ConfigValidate:0x00000003894d80>
ERROR> INFO warden: Calling OUT action: #<VagrantPlugins::ProviderVirtualBox::Action::CheckVirtualbox:0x00000003894da8>
ERROR> INFO warden: Calling OUT action: #<Proc:0x00000003b13048@/opt/vagrant/embedded/gems/gems/vagrant-1.8.4/lib/vagrant/action/warden.rb:94 (lambda)>
ERROR> INFO warden: Calling OUT action: #<VagrantPlugins::ProviderVirtualBox::Action::MatchMACAddress:0x00000003b13098>
ERROR> INFO warden: Calling OUT action: #<VagrantPlugins::ProviderVirtualBox::Action::DiscardState:0x00000003b130c0>
ERROR> INFO warden: Calling OUT action: #<VagrantPlugins::ProviderVirtualBox::Action::Import:0x00000003b130e8>
ERROR> INFO warden: Calling OUT action: #<VagrantPlugins::ProviderVirtualBox::Action::PrepareCloneSnapshot:0x00000003608cb0>
ERROR> INFO warden: Calling OUT action: #<Vagrant::Action::Builtin::PrepareClone:0x00000001e50ed8>
ERROR> INFO warden: Calling OUT action: #<VagrantPlugins::ProviderVirtualBox::Action::Customize:0x00000001e50f00>
ERROR> INFO warden: Calling OUT action: #<VagrantPlugins::ProviderVirtualBox::Action::CheckAccessible:0x00000001e50f28>
ERROR> INFO warden: Calling OUT action: #<Proc:0x00000003ae3280@/opt/vagrant/embedded/gems/gems/vagrant-1.8.4/lib/vagrant/action/warden.rb:94 (lambda)>
ERROR> INFO warden: Calling OUT action: #<Vagrant::Action::Builtin::Call:0x00000003894df8>
ERROR> INFO warden: Calling OUT action: #<Vagrant::Action::Builtin::ConfigValidate:0x00000003894e20>
ERROR> INFO warden: Calling OUT action: #<Proc:0x00000002c0a088@/opt/vagrant/embedded/gems/gems/vagrant-1.8.4/lib/vagrant/action/warden.rb:94 (lambda)>
ERROR> INFO warden: Calling OUT action: #<Vagrant::Action::Builtin::HandleBox:0x00000002c45bb0>
ERROR> INFO warden: Calling OUT action: #<Proc:0x00000002b12fb8@/opt/vagrant/embedded/gems/gems/vagrant-1.8.4/lib/vagrant/action/warden.rb:94 (lambda)>
ERROR> INFO warden: Calling OUT action: #<Vagrant::Action::Builtin::Call:0x00000003894e70>
ERROR> INFO warden: Calling OUT action: #<VagrantPlugins::ProviderVirtualBox::Action::CheckVirtualbox:0x00000003894e98>
ERROR> INFO interface: Machine: action ["up", "end", {:target=>:default}]
ERROR> INFO environment: Released process lock: machine-action-23094bb78005cd19e7080392623f720c
ERROR>DEBUG environment: Attempting to acquire process-lock: dotlock
ERROR> INFO environment: Acquired process lock: dotlock
ERROR> INFO environment: Released process lock: dotlock
ERROR> INFO environment: Running hook: environment_unload
ERROR> INFO runner: Preparing hooks for middleware sequence...
ERROR> INFO runner: 2 hooks defined.
ERROR> INFO runner: Running action: environment_unload #<Vagrant::Action::Builder:0x00000003a06010>
waitFor returned:0
output thread died
error thread died
```

File: lib/landrush/server.rb (hunk 2 above)
Comment (dustymabe) [inline lib/landrush/server.rb]: +1

**Comment by dustymabe** [comment]
LGTM - but note I did not test this myself, so it's worth $.02

**Comment by bexelbie** [comment]
merged.

**Pull request closed without merging**